### ISSUE TITLE
How can runtimes help bundle authors with dependencies?

### ISSUE THREAD

**Issue by ryanmoran** [cnabio/cnab-spec#156]
There is already an outstanding [issue](https://github.com/deislabs/cnab-spec/issues/132) and [PR](https://github.com/deislabs/cnab-spec/pull/133) that enables bundles to share configuration data. In addition to this configuration concern, I think there is also a question of how bundles might declare further runtime dependencies.

As an example, imagine that I am building an application bundle that depends upon a separate database bundle. The current thinking around `outputs` would cover much of my current concerns for how my database bundle might share configuration with my application bundle.

#### Database Bundle `outputs`
```json
{
  "outputs": {
    "address": {
      "type": "string",
      "path": "/cnab/app/outputs/address"
    },
    "port": {
      "type": "integer",
      "path": "/cnab/app/outputs/port"
    },
    "name": {
      "type": "string",
      "path": "/cnab/app/outputs/name"
    },
    "username": {
      "type": "string",
      "path": "/cnab/app/outputs/username"
    },
    "password": {
      "type": "string",
      "path": "/cnab/app/outputs/password"
    }
  }
}
```

#### Application Bundle `parameters`
```json
{
  "parameters": {
    "address": {
      "type": "string"
    },
    "port": {
      "type": "integer"
    },
    "name": {
      "type": "string"
    },
    "username": {
      "type": "string"
    },
    "password": {
      "type": "string"
    }
  }
}
```

But this implementation doesn't really enforce a particular dependency of my application, which is that it requires a PostgreSQL database, and will break horribly if I accidentally configure it with MySQL.

Maybe I could add an `adapter` parameter to my bundle like the following:
```json
{
  "parameters": {
    "address": {
      "type": "string"
    },
    "port": {
      "type": "integer"
    },
    "name": {
      "type": "string"
    },
    "username": {
      "type": "string"
    },
    "password": {
      "type": "string"
    },
    "adapter": {
      "type": "string",
      "const": "postgres"
    }
  }
}
```

I can then enforce this dependency if the database bundle author then includes an output like the following:
```json
{
  "outputs": {
    "address": {
      "type": "string",
      "path": "/cnab/app/outputs/address"
    },
    "port": {
      "type": "integer",
      "path": "/cnab/app/outputs/port"
    },
    "name": {
      "type": "string",
      "path": "/cnab/app/outputs/name"
    },
    "username": {
      "type": "string",
      "path": "/cnab/app/outputs/username"
    },
    "password": {
      "type": "string",
      "path": "/cnab/app/outputs/password"
    },
    "adapter": {
      "type": "string",
      "default": "postgres",
      "path": "/cnab/app/outputs/adapter"
    }
  }
}
```

But maybe I can only run on version 10 and higher. Things quickly start to get kind of complicated. I could rely on an output in the database bundle like the `version` field shown below.
```json
{
  "outputs": {
    "address": {
      "type": "string",
      "path": "/cnab/app/outputs/address"
    },
    "port": {
      "type": "integer",
      "path": "/cnab/app/outputs/port"
    },
    "name": {
      "type": "string",
      "path": "/cnab/app/outputs/name"
    },
    "username": {
      "type": "string",
      "path": "/cnab/app/outputs/username"
    },
    "password": {
      "type": "string",
      "path": "/cnab/app/outputs/password"
    },
    "adapter": {
      "type": "string",
      "default": "postgres",
      "path": "/cnab/app/outputs/adapter"
    },
    "version": {
      "type": "string",
      "default": "10.7",
      "path": "/cnab/app/outputs/version"
    }
  }
}
```

How do I declare a dependency on PostgreSQL v10 and higher? Maybe a regex? Maybe some kind of custom "version" `format` field (this requires a runtime that enforces that format).

How are folks imagining exposing these types of dependency details that aren't specifically core to configuration, but have major impacts on the running service deployed by a bundle? We are happy to experiment with some ideas in a `custom` field for our own bundles, but wanted to see what folks thoughts were here too.

cc/ @youreddy @jeremyrickard @carolynvs

**Comment by carolynvs** [comment]
@ryanmoran It looks like you are overloading dependency management, being able to swap out dependent bundles based on if it satisfies a set of requirements (and it may not just be as simple as single version range and name of a bundle), onto the existing CNAB parameter construct. 

IMO if that's what you are looking for, we maybe should explicitly call it out as such "Dependency Management / Resolution" and consider if it should have first class support in the spec. Or at least let's chat about it some more over zoom?

This is something that I'd be interested in discussing further as it's a problem that every client with dependency support is going to try to solve, and if it doesn't end up in the spec proper, it is going to be unreliable/unportable with informal "oh yeah we just shove version in the params and hope we can parse it out later". 😀 

Err, so yeah 💯 from me. This needs to happen but I'm not sure yet that we should do it by reusing parameters.

**Comment by carolynvs** [comment]
Sorry I hate it when people say "don't do that" and then don't provide a suggestion for what to do! 😊 

What I meant by that was that if we rely on each author or client's own idea of what to name the parameter for how to filter "I want mysql version X" then each will pick something different. In order for this to work well, it's important that all the bundle authors are using the same terms/mechanism to state their requirements so that we don't end up with a splintered set of bundles where some use "postgres-version" and others use "db-version" or something like that.

We could very easily use parameters, and then legislate in the 800's section of the spec on what names to use. Though I do feel that perhaps if the bundle authors aren't actually needing those parameters that we are reusing them just because they are there? To me this type of information is important metadata about the bundle which is why you immediately wanted to make those parameters constants, which tells me that they aren't really parameters. 

I would be interested in discussing making this information proper supporter fields in the bundle, perhaps in the metadata as well-defined fields that clients that support dependency management know to look for? This is a unvetted example of the idea (putting the information in the metadata instead of as a set of properties):

```
"metadata:" { 
  "requires": [
    { "type": "postgres", "app-version": "10" }
  ]
}
```

Then in the spec having a dependency management sub-spec(?) that says if you support dependency management, you understand how to read out the `requires` document in the metadata, and will use a particular schema to communicate what you require.

It would also need a recommended way or set of names that bundle authors agree to use when asking for "postgres" so that some people don't use "postgres", and others use "postgresql" and others "postgresdb", etc. That's getting into the details a bit, probably better left for a discussion but I wanted to bring that up as a known problem that I've run into before in this space.

**Comment by ryanmoran** [comment]
@carolynvs, I strongly agree with you that what I have above is a hack. In the example, I'm abusing `parameters` and `outputs` to get this dependency resolution concept, and its definitely not how those APIs were intended to be used.

I am also much more inclined to believe that what you are describing with the `provides/requires` API is the right path. I'd be happy with a PR that looked roughly like what you are proposing that would allow bundle authors to declare the "types" of services they install or depend upon.

As far as the version values, it would be nice to have some ability to declare a SemVer range dependency like the following: `>=10.2.2 <10.3.0`.

**Comment by carolynvs** [comment]
Unless you'd like to do the honors, I can submit a PR next week that looks like what I have in my comment and let the rotten fruit fly! 😉

**Comment by simonferquel** [comment]
I thought dependencies were out of scope for v1 of the spec.
I think the dependency problem could be easily resolved without specifying it, by making an opinated cnab building tool similar to porter, but aggregating CNABs instead of trying to build a single invocation image.

**Comment by carolynvs** [comment]
> I think the dependency problem could be easily resolved without specifying it

Oof, I'm sure you mean well but really that is minimizing what is a real problem that we care about and I would appreciate that we try to not say things like that in general. It makes discourages people from wanting to participate in a conversations when their concerns are dismissed like that.

Currently porter has a way for depending on other specific named bundles (which is much more limited than what Ryan is suggesting in this issue). Really that only solves the problem for just bundles that are authored by porter and creates a subset of bundles that can only depend on other porter bundles. I much prefer to foster an ecosystem that encourages growth, and continues to keep the tools working together. So that if a tool is "cnab compliant" it still mostly works with the other tools.

Since what we are proposing would be additive, I understood it would not be disallowed. If the group had previously decided that dependencies were off the table for v1 and should be held off on discussions for X amount of time, then I'd appreciate a link to that decision. I'm having trouble finding that in GitHub.

**Comment by simonferquel** [comment]
@carolynvs sorry if I sounded offensive, it was absolutely not my intent.
In the past I proposed something similar (I think it was on the old docker-hosted slack channel), and the conclusion to the discussion was that we'd prefer to evaluate how a CNAB composition model could work around this problem (in a classic "prefer composition over inheritance" way of thinking).

I think we did not (yet) work seriously enough on this issue, and I'd like to look at CNAB composition first before introducing bundle dependencies.

My main concern about dependencies, is about distribution: for now the CNAB spec focuses on runtime structures and we want separation of concern between those data structures, and the way bundles are distributed.
When(/if) we introduce bundle dependencies, we need to be very carefull: how do we represent the dependency in the runtime datastructure ? do introduce a dependency on how those bundle are distributed ? Or maybe we can have a different representation: the runtime datastructure could actually embed dependant CNABs ? but how do we author them by putting only a reference to a registry-stored CNAB then ? All those questions are clearly not straight-forward to answer, and if we could solve the issue by introducing a CNAB composition tool providing a good experience working around those issues, I would personnally feel better :).

**Comment by glyn** [comment]
Please note that the [OSGi](https://www.osgi.org/developer/specifications/) standard addresses a very similar inter-bundle [1] dependency requirement and has a [NP-complete dependency resolution algorithm](https://stackoverflow.com/questions/2085106/is-the-resolution-problem-in-osgi-np-complete). The OSGi resolver was a tricky thing to implement, so beware repeating the exercise. One solution is to avoid _optional_ dependencies.

1. Coincidentally, OSGi also packages software as _bundles_

**Comment by carolynvs** [comment]
Thanks for the link! I ran into the same category of problem working on the solver for dep and ughhhhhh, agreed this is a problem that we want to avoid.

**Comment by technosophos** [comment]
Is there anything left to do on this? It appears to have been largely addressed by #166.
